### PR TITLE
Add Takeoff and Landing Commands

### PR DESCRIPTION
**Pull request by wuda17**
added `add_takeoff_and_landing_command.py` into modules and unit test `test_add_takeoff_and_landing_command.py`

### CODE REVIEW

File: modules/add_takeoff_and_landing_command.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+Prefixes a takeoff command and suffixes a landing command to the end of the list of commands.
+"""
+
+import dronekit
+
+
+MAVLINK_TAKEOFF_FRAME = dronekit.mavutil.mavlink.MAV_FRAME_GLOBAL_RELATIVE_ALT
+MAVLINK_LANDING_FRAME = dronekit.mavutil.mavlink.MAV_FRAME_GLOBAL
+MAVLINK_TAKEOFF_COMMAND = dronekit.mavutil.mavlink.MAV_CMD_NAV_RETURN_TO_LAUNCH
+MAVLINK_LANDING_COMMAND = dronekit.mavutil.mavlink.MAV_CMD_NAV_TAKEOFF
```
Comment (Xierumeng) [inline modules/add_takeoff_and_landing_command.py:11]: Swapped?

File: modules/add_takeoff_and_landing_command.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+def add_takeoff_and_landing_command(commands: "list[dronekit.Command]",
+                                    altitude: float) -> "list[dronekit.Command]":
+    """
+    Prepends a takeoff command and appends a landing command to a list of dronekit commands
+
+    Parameters
+    ----------
+    commands: list[dronekit.Command]
+        dronekit commands that can be sent to the drone.
```
Comment (Xierumeng) [inline modules/add_takeoff_and_landing_command.py:22]: Capitalize 1st word. Same with the other descriptions.

File: tests/test_add_takeoff_and_landing_command.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+from modules import add_takeoff_and_landing_command
+
+
+def test_add_takeoff_and_landing_command():
+    """
+    Tests functionality correctness of add_takeoff_and_landing_command
+    """
+    commands = []
+    altitude = 50
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:15]: Make this global constant.

File: tests/test_add_takeoff_and_landing_command.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    # Test takeoff command
+    takeoff_command = commands_actual[0]
+    assert isinstance(takeoff_command, dronekit.Command)
+    assert takeoff_command.frame == add_takeoff_and_landing_command.MAVLINK_TAKEOFF_FRAME
+    assert takeoff_command.command == add_takeoff_and_landing_command.MAVLINK_TAKEOFF_COMMAND
+    assert takeoff_command.param7 == altitude
+
+    # Test landing command
+    landing_command = commands_actual[1]
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:30]: Use index -1 for the element at the end of the list.

File: tests/test_add_takeoff_and_landing_command.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    assert isinstance(landing_command, dronekit.Command)
+    assert landing_command.frame == add_takeoff_and_landing_command.MAVLINK_LANDING_FRAME
+    assert landing_command.command == add_takeoff_and_landing_command.MAVLINK_LANDING_COMMAND
+
+    # Ensure the commands list is unmodified
+    assert len(commands) == 0
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:36]: Does this pass? I wouldn't think so because Python passes everything by reference. Explanation: https://stackoverflow.com/questions/986006/how-do-i-pass-a-variable-by-reference

File: tests/test_add_takeoff_and_landing_command.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import dronekit
+
+from modules import add_takeoff_and_landing_command
+
+
+def test_add_takeoff_and_landing_command():
+    """
+    Tests functionality correctness of add_takeoff_and_landing_command
+    """
+    commands = []
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:14]: Add a test for non-empty list. This is where Pytest fixtures would be useful, as well as a generalized function with assertions that is called by the test functions.

File: tests/test_add_takeoff_and_landing_command.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            0,
+            0,
+            0,
+        ),
+        dronekit.Command(
+            0,
+            0,
+            0,
+            MAVLINK_TEST_FRAME,
+            MAVLINK_TEST_FRAME,
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:43]: ?

File: tests/test_add_takeoff_and_landing_command.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            0,  # param1
+            0,
+            0,
+            0,
+            0,
+            0,
+            0,
+        )
+    ]
+    return commands
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:55]: `yield`

File: tests/test_add_takeoff_and_landing_command.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+
+@pytest.fixture
+def empty_commands():
+    """
+    Fixture for an empty list of commands.
+    """
+    commands = []
+    return commands
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:64]: `yield`

File: tests/test_add_takeoff_and_landing_command.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+            MAVLINK_TEST_COMMAND,
+            0,
+            0,
+            0,  # param1
+            0,
+            0,
+            0,
+            0,
+            0,
+            0,
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:36]: Not sure if all zeroes here is allowed by this particular MAVLink.

File: tests/test_add_takeoff_and_landing_command.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+            MAVLINK_TEST_FRAME,
+            0,
+            0,
+            0,  # param1
+            0,
+            0,
+            0,
+            0,
+            0,
+            0,
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:52]: Not sure if all zeroes here is allowed by this particular MAVLink.

File: tests/test_add_takeoff_and_landing_command.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+@pytest.fixture
+def empty_commands():
+    """
+    Fixture for an empty list of commands.
+    """
+    commands = []
+    return commands
+
+
+def assert_takeoff_and_landing_commands(commands_input, commands_actual, altitude):
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:67]: Comparison order is actual then expected (e.g. `assert actual == expected` ), change signature to follow this order.

Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:67]: Add type annotation.

Comment (Xierumeng) [reply tests/test_add_takeoff_and_landing_command.py:67]: Actually, remove `commands_input` if it's not being used.

File: tests/test_add_takeoff_and_landing_command.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    assert takeoff_command.z == altitude
+
+    # Test landing command
+    landing_command = commands_actual[-1]
+    assert isinstance(landing_command, dronekit.Command)
+    assert landing_command.frame == add_takeoff_and_landing_command.MAVLINK_LANDING_FRAME
+    assert landing_command.command == add_takeoff_and_landing_command.MAVLINK_LANDING_COMMAND
+
+
+def test_add_takeoff_and_landing_on_empty_commands(empty_commands):
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:87]: Type annotation.

File: tests/test_add_takeoff_and_landing_command.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    """
+    Tests functionality correctness of add_takeoff_and_landing_command on empty list of commands.
+    """
+    commands_actual = add_takeoff_and_landing_command.add_takeoff_and_landing_command(empty_commands, ALTITUDE)
+    assert_takeoff_and_landing_commands(empty_commands, commands_actual, ALTITUDE)
+
+
+def test_add_takeoff_and_landing_on_nonempty_commands(non_empty_commands):
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:95]: Type annotation.

File: tests/test_add_takeoff_and_landing_command.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    assert isinstance(landing_command, dronekit.Command)
+    assert landing_command.frame == add_takeoff_and_landing_command.MAVLINK_LANDING_FRAME
+    assert landing_command.command == add_takeoff_and_landing_command.MAVLINK_LANDING_COMMAND
+
+
+def test_add_takeoff_and_landing_on_empty_commands(empty_commands):
+    """
+    Tests functionality correctness of add_takeoff_and_landing_command on empty list of commands.
+    """
+    commands_actual = add_takeoff_and_landing_command.add_takeoff_and_landing_command(empty_commands, ALTITUDE)
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:91]: You can do a deep copy of the list so that 1 goes here (and is mutated) and 1 is used for comparison.

File: tests/test_add_takeoff_and_landing_command.py [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+            MAVLINK_TEST_FRAME,
+            MAVLINK_TEST_COMMAND,
+            0,
+            0,
+            0,  # param1
+            ACCEPT_RADIUS,
+            0,
+            0,
+            42.123,
+            -73.456,
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:37]: Use constants.

File: tests/test_add_takeoff_and_landing_command.py [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+            MAVLINK_TEST_FRAME,
+            MAVLINK_TEST_COMMAND,
+            0,
+            0,
+            0,  # param1
+            ACCEPT_RADIUS,
+            0,
+            0,
+            42.789,
+            -73.987,
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:53]: Use constants.

File: tests/test_add_takeoff_and_landing_command.py [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    assert_expected_takeoff_and_landing_commands(commands_actual, commands_expected, ALTITUDE)
+
+
+def test_add_takeoff_and_landing_on_nonempty_commands(non_empty_commands: "list[dronekit.Command]"):
+    """
+    Tests functionality correctness of add_takeoff_and_landing_command on non-empty list of commands.
+    """
+    commands_expected = copy.deepcopy(non_empty_commands)
+    commands_actual = add_takeoff_and_landing_command.add_takeoff_and_landing_command(non_empty_commands, ALTITUDE)
+    assert_expected_takeoff_and_landing_commands(commands_actual, commands_expected, ALTITUDE)
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:111]: Nit: Add single empty line at end of file.

File: tests/test_add_takeoff_and_landing_command.py [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+    # Test landing command
+    landing_command = commands_actual[-1]
+    assert isinstance(landing_command, dronekit.Command)
+    assert landing_command.frame == add_takeoff_and_landing_command.MAVLINK_LANDING_FRAME
+    assert landing_command.command == add_takeoff_and_landing_command.MAVLINK_LANDING_COMMAND
+
+    # Test original commands
+    assert len(commands_actual) == len(commands_expected) + 2
+    for i in range(len(commands_expected)):
+        assert commands_actual[i+1] == commands_expected[i]
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:93]: Try slicing `commands_actual` and doing a list comparison: https://stackoverflow.com/a/69923697

**Review by Xierumeng — APPROVED** [review]
